Use contiguous recurrence for c-related sum computations

TSGut opened the pull request:
@ioannisPApapadopoulos @dlfivefifty 

Basically just porting over @MikaelSlevinsky's contiguous recurrence implementation in a way that should be reasonably natural for this package. This should be both much faster and hopefully much more accurate at high c.

Previously we had
```julia
julia> Wcomp = SemiclassicalJacobiWeight.(1.1,2:2,3:3,3:100);

julia> @time scomp = sum.(Wcomp);
  0.112850 seconds (336.34 k allocations: 14.352 MiB)
```
Now we get
```julia
julia> W = SemiclassicalJacobiWeight.(1.1,2,3,3:100);

julia> @time s = sum.(W);
  0.000048 seconds (28 allocations: 2.156 KiB)
```
At no cost to accuracy (the old method using bigfloat, the new one just Float64):
```julia
julia> s - scomp
98-element Vector{Float64}:
 -2.6020852139652106e-18
 -2.6020852139652106e-18
 -1.734723475976807e-18
 -1.3010426069826053e-18
 -8.673617379884035e-19
  0.0
 -4.336808689942018e-19
 -2.168404344971009e-19
 -2.168404344971009e-19
  ⋮
  5.898059818321144e-17
  5.898059818321144e-17
  6.591949208711867e-17
  7.28583859910259e-17
  7.632783294297951e-17
  7.979727989493313e-17
  8.673617379884035e-17
  9.367506770274758e-17
  1.0408340855860843e-16
```

It's implemented to work whenever sum is broadcast over a crange vector of weights as seen above, as it is it wont be used if you just call sum on a single weight (mainly because doing that in a clean way would involve a lot of clauses about a, b and c and other contiguous recurrences), maybe we implement that at some point in the future. If you must use it on a single weight just do e.g. `sum.(SemiclassicalJacobiWeight.(1.1,2,3,3:3))`.

Still need to add and run tests.

TSGut commented:
Now also featuring the weighted conversion speedup promised to @ioannisPApapadopoulos. Hopefully this getting faster and faster eventually reaches a point where it is fast enough (I suspect now it would just be code optimization rather than better formulas tho).

Here is the example from https://github.com/JuliaApproximation/SemiclassicalOrthogonalPolynomials.jl/issues/78 reran in current implementation:

```julia
julia> using SemiclassicalOrthogonalPolynomials, ClassicalOrthogonalPolynomials

julia> t = 1.1
1.1

julia> Q₀₀ = SemiclassicalJacobi.(t, 0, 0, 0:∞);

julia> Q₁₁ = SemiclassicalJacobi.(t, 1, 1, 0:∞);

julia> L  = (Weighted.(Q₀₀) .\ Weighted.(Q₁₁));

julia> @time L[20]; # before: 0.183541 seconds
  0.138117 seconds (237.30 k allocations: 11.231 MiB, 60.01% compilation time: 100% of which was recompilation)

julia> using SemiclassicalOrthogonalPolynomials, ClassicalOrthogonalPolynomials

julia> t = 1.1
1.1

julia> Q₀₀ = SemiclassicalJacobi.(t, 0, 0, 0:∞);

julia> Q₁₁ = SemiclassicalJacobi.(t, 1, 1, 0:∞);

julia> L  = (Weighted.(Q₀₀) .\ Weighted.(Q₁₁));

julia> @time L[20]; # before: 0.183541 seconds
  0.060309 seconds (191.13 k allocations: 8.075 MiB)

julia> @time L[30]; # before: 0.396986 seconds
  0.100929 seconds (402.88 k allocations: 15.625 MiB)

julia> @time L[40]; # before: 0.861748 seconds
  0.212241 seconds (758.47 k allocations: 27.565 MiB)

julia> @time L[50]; # before: 1.565395 seconds
  0.291458 seconds (1.26 M allocations: 43.963 MiB)

julia> @time L[60]; # before: 6.235148 seconds (with compile), 2.743544 seconds (without compile)
  0.441863 seconds (1.95 M allocations: 64.999 MiB)
```

and we also get 

```julia
julia> @time L[250]
 37.124558 seconds (142.60 M allocations: 3.696 GiB, 2.71% gc time)
(((ℵ₀×ℵ₀ LinearAlgebra.LowerTriangular{Float64, LinearAlgebra.Adjoint{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrices.BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, LazyBandedMatrices.SymTridiagonal{Float64, LazyArrays.BroadcastVector{Float64, typeof(-), Tuple{Bool, SubArray{Float64, 1, ClassicalOrthogonalPolynomials.QRJacobiData{:Q, Float64}, Tuple{Base.Slice{InfiniteArrays.OneToInf{Int64}}, Int64}, false}}}, LazyArrays.BroadcastVector{Float64, typeof(-), Tuple{SubArray{Float64, 1, ClassicalOrthogonalPolynomials.QRJacobiData{:Q, Float64}, Tuple{Base.Slice{InfiniteArrays.OneToInf{Int64}}, Int64}, false}}}}}}} with indices OneToInf()×OneToInf()) * (ℵ₀×ℵ₀ LinearAlgebra.Diagonal{Float64, FillArrays.Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}} with indices OneToInf()×OneToInf()) with indices OneToInf()×OneToInf()) * ((ℵ₀×ℵ₀ LinearAlgebra.LowerTriangular{Float64, LinearAlgebra.Adjoint{Float64, InfiniteLinearAlgebra.AdaptiveCholeskyFactors{Float64, BandedMatrices.BandedMatrix{Float64, Matrix{Float64}, Base.OneTo{Int64}}, LazyBandedMatrices.SymTridiagonal{Float64, SubArray{Float64, 1, ClassicalOrthogonalPolynomials.QRJacobiData{:Q, Float64}, Tuple{Base.Slice{InfiniteArrays.OneToInf{Int64}}, Int64}, false}, SubArray{Float64, 1, ClassicalOrthogonalPolynomials.QRJacobiData{:Q, Float64}, Tuple{Base.Slice{InfiniteArrays.OneToInf{Int64}}, Int64}, false}}}}} with indices OneToInf()×OneToInf()) * (ℵ₀×ℵ₀ LinearAlgebra.Diagonal{Float64, FillArrays.Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}} with indices OneToInf()×OneToInf()) with indices OneToInf()×OneToInf()) with indices OneToInf()×OneToInf()) * (ℵ₀×ℵ₀ LinearAlgebra.Diagonal{Float64, FillArrays.Fill{Float64, 1, Tuple{InfiniteArrays.OneToInf{Int64}}}} with indices OneToInf()×OneToInf()) with indices OneToInf()×OneToInf():
  0.00434421    ⋅            ⋅             ⋅           …  
  0.00428913   0.00608072    ⋅             ⋅
 -3.76572e-5   0.0059776    0.00737181     ⋅
   ⋅          -7.88228e-5   0.00721544    0.00842674
   ⋅            ⋅          -0.000127024   0.00821237
   ⋅            ⋅            ⋅           -0.000180951  …
   ⋅            ⋅            ⋅             ⋅
   ⋅            ⋅            ⋅             ⋅
   ⋅            ⋅            ⋅             ⋅
  ⋮                                                    ⋱
```

Granted that is not exactly fast but there are a LOT of cholesky decompositions that need to be done to construct this.

TSGut commented:
Ok I think this does what I wanted it to now. @ioannisPApapadopoulos give it a try and see if this fixes the issue you saw with the normalization. You should precompute the normalization constants if possible so it can all be done in one go of the recurrence.

ioannisPApapadopoulos commented:
Brilliant, thanks @TSGut. I'll give it a spin and report back.

dlfivefifty commented:
I don't really understand what `L[250]` is computing. Why is it quadratic in cost? I would think computing `L[m][1:n,1:n]` would be `O(m*n)` operations

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -506,6 +506,50 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+# here we construct hierarchies of c weight sums by means of contiguous recurrence relations
+###
+
+mutable struct SemiclassicalJacobiCWeightFamily{T, C} <: AbstractVector{SemiclassicalJacobiWeight{T}}
```
This time is redundant, isn't it the same as `SemiclassicalJacobiFamily{T, T, T, C}`?

TSGut replied:
I am confused, how is it the same? This is a vector containing weights with a range in the C index, the other is a vector of the polynomial bases themselves

TSGut replied:
the goal is to compute `sum.(::Vector{SemiclassicalJacobiWeight})` using the contiguous recurrence in a single run when the entries only differ in the c parameter.

dlfivefifty commented on test/runtests.jl:
```diff
@@ -354,8 +354,7 @@ end
     @testset "conversion" begin
-        D = legendre(0..1) \ P
-        @test (P \ legendre(0..1))[1:10,1:10] == inv(Matrix(D[1:10,1:10]))
+        @test_broken (P \ legendre(0..1))[1:10,1:10] == inv(Matrix((legendre(0..1) \ P)[1:10,1:10]))
```
What broke?

TSGut replied:
Yeah good question. this is on my todo list to investigate for tomorrow / this week but basically it thinks some definition is ambiguous in the newly broken line. I suspect this is unrelated to changes I made (another I think unrelated test broke due to ambiguity which I already fixed). It probably just needs to be redirected to the correct of the two ambiguous definitions and it will work again

TSGut commented:
> I don't really understand what `L[250]` is computing. Why is it quadratic in cost? I would think computing `L[m][1:n,1:n]` would be `O(m*n)` operations

A fair question. In theory, the way this is computed is by chaining lots of QR or Cholesky decompositions to ladder up. But the cost is not straightforward because this laddering involves some cached arrays meaning that

```julia
julia> @time L[250];
 32.742888 seconds (140.99 M allocations: 3.648 GiB, 2.50% gc time)

julia> @time L[250];
 16.266950 seconds (72.01 M allocations: 1.871 GiB, 2.54% gc time)
```
So at least half of the computing time here is spent filling in some cached jacobimatrices to then perform cholesky decomposition on.

TSGut commented:
I think I know where most of the computation time is lost, possibly what causes the N^2 cost. I will see if I can fix it.

TSGut commented:
hm, ok it is faster now, especially after caching which is nice but still has not changed the complexity.

```julia
julia> @time LL[250];
 25.778077 seconds (106.66 M allocations: 2.762 GiB, 2.96% gc time, 0.09% compilation time)

julia> @time LL[250];
  8.797171 seconds (36.06 M allocations: 959.496 MiB, 2.45% gc time)
```

TSGut commented:
Ok this is more like it... 

```julia
julia> @time L[250];
 16.329566 seconds (70.79 M allocations: 1.829 GiB, 2.53% gc time)

julia> @time L[250];
  0.000901 seconds (835 allocations: 39.078 KiB)
```

Importantly I think this reveals why this is taking longer than maybe expected. It precomputes all the previous stuff too on the way

```julia
julia> @time L[250];
  0.000896 seconds (835 allocations: 39.078 KiB)

julia> @time L[249];
  0.074969 seconds (309.57 k allocations: 7.869 MiB)

julia> @time L[249];
  0.000951 seconds (835 allocations: 39.078 KiB)
```

@ioannisPApapadopoulos you should keep this in mind when using this. Probably don't call L[1:250], call L[250] on its own first otherwise Julias ordering may not make use of this.

dlfivefifty commented:
If it’s ` AbstractCachedVector ` then `   L[1:250] ` should be fine

TSGut commented:
retagged as WIP for now because of a potential forward recurrence instability. see the relevant issue in the draft repo

TSGut commented:
Now includes `sumquotient()` used in weighted conversions to avoid the `t^c` term error in computations of `sum(wP)/sum(wQ)` by cancelling them before using the recurrence.

TSGut commented:
While there are some drawbacks here that we discussed, I think we probably still merge this in this form (after resolving conflicts)? The previous implementation is pretty much strictly worse so we can use the recurrence until we revamp the decompositions stuff to keep track of the normalization directly.

ioannisPApapadopoulos commented:
I think there is a bug when `c=0`. e.g.

```
using SemiclassicalOrthogonalPolynomials
import HypergeometricFunctions: _₂F₁general2
using SpecialFunctions

t, a, b, c = 1.1, 1, 1, 0
sum.(SemiclassicalJacobiWeight.(t,a,b,c:c)) # returns 1.0
t^c * beta(1+a,1+b) * _₂F₁general2(1+a,-c,2+a+b,1/t) # returns 0.16666666666666669
```

TSGut commented:
Yep, that's definitely a bug. Will fix.

ioannisPApapadopoulos commented:
@TSGut have you made a start? or should I have a go fixing it next week?

TSGut commented:
Ah, sorry I just didn't push. It was just a bug in the special casing for the c=0.

TSGut commented:
@dlfivefifty ready for review or merge

ioannisPApapadopoulos commented:
Thank you!

dlfivefifty commented:
Generally you would resolve the conflicts and have tests passing before asking someone to review or merge

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -48,9 +48,9 @@ function getindex(P::SemiclassicalJacobiWeight, x::Real)
-    # (t,a,b,c) = P.t, P.a, P.b, P.c
-    return convert(T, t^c * beta(1+a,1+b) * _₂F₁general2(1+a,-c,2+a+b,1/t))
+    (t,a,b,c) = (P.t, P.a, P.b, P.c)
+    t,a,b,c = BigFloat("$t"),BigFloat("$a"),BigFloat("$b"),BigFloat("$c") # This is needed at high parameter values
```
why would you do `BigFloat("$t")`? that is crazy. Use `convert(BigFloat, t)`

TSGut replied:
We can do what you suggest but it loses some accuracy without really affecting speed too much.

TSGut replied:
Note

```julia
julia> t = 1.1
1.1

julia> BigFloat("1.1")-convert(BigFloat,t)
-8.881784197001252323389053344726562499999999999999999999999999654553257796222215e-17

julia> BigFloat("1.1")-BigFloat("$t")
0.0
```

and this really matters here.

dlfivefifty replied:
This is nonsense. Here t is not 1.1 but fl(1.1) ie the rounded version. The fact it happens to print out as 1.1 is a coincidence.

dlfivefifty replied:
So what you are doing is losing accuracy because it’s rounding a base-2 number to base-10

TSGut replied:
Well, it is only coincidence if we ignore that t is a user input number but I don't mind changing it, the truth is that if a user, i.e. we, really properly needs accuracy on these they should construct the polynomials with BigFloat in the first place.

dlfivefifty replied:
How on earth would we know the user had originally typed in 1.1?

And your approach is introducing error extra error if t = 1/3. That is you round to base 2 and then round again to base 10, then round to base 2 with higher precision.

TSGut replied:
fixed

TSGut commented:
> Generally you would resolve the conflicts and have tests passing before asking someone to review or merge

Of course. I just missed that there was a conflict, sorry about that.

TSGut commented:
@dlfivefifty Ok, all passing and conflicts and review resolved.

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -509,6 +512,79 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+# here we construct hierarchies of c weight sums by means of contiguous recurrence relations
+###
+
+mutable struct SemiclassicalJacobiCWeightFamily{T, C} <: AbstractVector{SemiclassicalJacobiWeight{T}}
```
Add a docstring: I have no idea what this is meant to be doing.

Also why is it mutable?

TSGut replied:
added a docstring, though it is intended to be an internal method that picks up on a computation important for annulus applications, not something that a user manually constructs. it is automatically used when broadcasting weight functions over the right kind of input. Won't argue with more docs though

TSGut replied:
the point is really that sum.(::SemiclassicalJacobiCWeightFamily) is fast and uses recurrence instead of individual sums via pFq

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -509,6 +512,79 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+    return abs.(convert.(T,t.^c.*exp(loggamma(a+1)+loggamma(b+1)-loggamma(a+b+2)).*getindex(F,W.c.+1)))
+end
+
+# computes sum(wP)/sum(wQ) by taking into account cancellations
```
Just make this comment a docstring so you can see it with `?`

TSGut replied:
done

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -509,6 +512,79 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+    # This is needed at high parameter values.
+    # Manually setting setprecision(2048) allows accurate computation even for very high c. 
+    t,a,b = convert(BigFloat,t),convert(BigFloat,a),convert(BigFloat,b)
+    sumw = (a,b,c,t) -> pFq((a+1,-c),(a+b+2, ), 1/t)
```
Why is this an anonymous function inside the function?

TSGut replied:
changed \ fixed

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -509,6 +512,79 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+    a = wP.a; b = wP.b; c = Int(wP.c); t = wP.t;
+    # This is needed at high parameter values.
+    t,a,b = convert(BigFloat,t),convert(BigFloat,a),convert(BigFloat,b)
+    sumw = (a,b,c,t) -> pFq((a+1,-c),(a+b+2, ), 1/t)
```
This is the same anonymous function repeated

TSGut replied:
changed \ fixed

dlfivefifty commented on src/SemiclassicalOrthogonalPolynomials.jl:
```diff
@@ -509,6 +512,79 @@ function LazyArrays.cache_filldata!(P::SemiclassicalJacobiFamily, inds::Abstract
+    F[1] = sumw(a,b,0,t) # c=0
+    cmax == 0 && return abs.(convert.(T,t.^c.*exp(loggamma(a+1)+loggamma(b+1)-loggamma(a+b+2)).*getindex(F,1:1)))
+    F[2] = sumw(a,b,1,t) # c=1
+    @inbounds for n in 1:cmax-1
```
Can't you just use `RecurrenceArray` for this? It would probably avoid the need for `BigFloat` all together

TSGut replied:
I haven't used RecurrenceArray in practice before, in theory from what you have explained it should be helpful here. I will give it a try and see if it works here.

TSGut commented:
Ok other than the recurrence array change I have addressed the reviews. I will check how much work it would be to convert this all to use RecurrenceArrays in the coming few days.
